### PR TITLE
[SPARK-8552] [THRIFTSERVER] Using incorrect database in multiple sessions

### PR DESCRIPTION
**Pull request by navis**
Currently, metaHive in HiveContext shares single SessionState instance with all execution threads, which makes problems especially with "use" command. I've also went into this problem and spent hard time to know what has happened. This is just first try and need more tests.

### CODE REVIEW

**Comment by marmbrus** [comment]
ok to test

**Comment by WangTaoTheTonic** [comment]
I tested this patch, but got this error when executing "show databases;" using beeline,

>>15/09/10 15:11:02 INFO SessionState: Created HDFS directory: /tmp/hive/root/fc5c8bbe-0e63-49f1-8286-ec51c4432b94/_tmp_space.db
15/09/10 15:11:02 INFO HiveSessionImpl: Operation log session directory is created: /tmp/root/operation_logs/fc5c8bbe-0e63-49f1-8286-ec51c4432b94
15/09/10 15:11:26 ERROR SparkExecuteStatementOperation: Error running hive query as user : root
java.lang.NullPointerException
        at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$1$$anon$2.run(SparkExecuteStatementOperation.scala:164)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:415)
        at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1628)
        at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation$$anon$1.run(SparkExecuteStatementOperation.scala:182)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
        at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
        at java.util.concurrent.FutureTask.run(FutureTask.java:166)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603)
        at java.lang.Thread.run(Thread.java:722)

**Comment by navis** [comment]
@WangTaoTheTonic I've also seen that rebasing this to spark-1.5.0. Seemed done something wrong in rebase process. I'll update soon.

**Comment by WangTaoTheTonic** [comment]
@navis Thanks you for the fix. I have tested "use $database" on local Thrift Server and the function is ok. 

There might still have a test failed need to fix.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientWrapper.scala [hunk 1]
```diff
@@ -57,10 +58,9 @@ import org.apache.spark.util.{CircularBuffer, Utils}
  * @param initClassLoader the classloader used when creating the `state` field of
```
Comment (WangTaoTheTonic) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientWrapper.scala]: The comments need to be updated with its construct args.

**Comment by navis** [comment]
@WangTaoTheTonic I've fixed test itself, which seemed invalid. "add jar" should be applied to the session which called it but in test, it expects added jar to be accessed with other sessions.

**Comment by WangTaoTheTonic** [comment]
@marmbrus How do you think this fix? As the issue priority is very high, I think we better fix it ASAP.

**Comment by marmbrus** [comment]
I think I prefer the solution taken in https://github.com/apache/spark/pull/8909, which limits the use of thread local variables

**Comment by davies** [comment]
Since #8909 is merged, would you mind close this PR?

**Comment by DoingDone9** [comment]
@navis 
when i use spark-sql and run sql like that 
"
add jar /home/udf-0.0.1-SNAPSHOT.jar;
create temporary function arr_greater_equal as 'com.xx.yy.dac.hive.udf.UDFArrayGreaterEqual';  
"
I got a error 
15/11/07 16:01:15 ERROR CliDriver: org.apache.spark.sql.execution.QueryExecutionException: FAILED: SemanticException [Error 10072]: Database does not exist: default
        at org.apache.spark.sql.hive.client.ClientWrapper$$anonfun$runHive$1.apply(ClientWrapper.scala:349)
        at org.apache.spark.sql.hive.client.ClientWrapper$$anonfun$runHive$1.apply(ClientWrapper.scala:326)
        at org.apache.spark.sql.hive.client.SharedWrapper.withHiveState(ClientWrapper.scala:469)
        at org.apache.spark.sql.hive.HiveContext$$anon$2.withHiveState(HiveContext.scala:194)
        at org.apache.spark.sql.hive.client.ClientWrapper.withHiveState(ClientWrapper.scala:138)
        at org.apache.spark.sql.hive.client.ClientWrapper.runHive(ClientWrapper.scala:326)
        at org.apache.spark.sql.hive.client.ClientWrapper.runSqlHive(ClientWrapper.scala:316)
        at org.apache.spark.sql.hive.HiveContext.runSqlHive(HiveContext.scala:542)
        at org.apache.spark.sql.hive.execution.HiveNativeCommand.run(HiveNativeCommand.scala:33)
        at org.apache.spark.sql.execution.ExecutedCommand.sideEffectResult$lzycompute(commands.scala:57)

**Pull request closed without merging**